infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/runtime/LogReplicationClientServerRouter.java

```diff
@@ -0,0 +1,605 @@
+import org.corfudb.runtime.proto.service.CorfuMessage;
+import org.corfudb.util.CFUtils;
+import org.corfudb.util.retry.IRetry;
+import org.corfudb.util.retry.IntervalRetry;
+import org.corfudb.util.retry.RetryNeededException;
+
+import javax.annotation.Nonnull;
+import java.io.File;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getUuidMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getDefaultProtocolVersionMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getRequestMsg;
+import static org.corfudb.protocols.service.CorfuProtocolMessage.getResponseMsg;
+
+/**
+ * A Base Source Router, which interfaces between the Log replication Source components (FSMs, AckReader, etc) and the
+ * custom transport adapter
+ *
+ * This class is inherited by both LogReplicationSourceClientRouter and LogReplicationSourceServerRouter.
+ */
+@Slf4j
+public class LogReplicationClientServerRouter implements IClientServerRouter {
+
+    public static String REMOTE_LEADER = "REMOTE_LEADER";
+
+    /**
+     * A map of remoteCluster to {@link CompletableFuture} for each session, which is completed when a connection,
+     * including a successful handshake, completes and messages can be sent
+     * to the remote node.
+     */
+    @Getter
+    private volatile Map<LogReplicationSession, CompletableFuture<Void>> sessionToLeaderConnectionFuture;
+
+    /**
+     * A map of session to the current requestID for the session.
+     */
+    @Getter
+    @SuppressWarnings("checkstyle:abbreviation")
+    private final Map<LogReplicationSession,AtomicLong> sessionToRequestIdCounter;
+
+    /**
+     * Sync call response timeout (milliseconds).
+     */
+    @Getter
+    @Setter
+    private long timeoutResponse;
+
+    /**
+     * The outstanding requests on this router.
+     */
+    @Getter
+    private final Map<LogReplicationSession, Map<Long, CompletableFuture>> sessionToOutstandingRequests;
+
+    /**
+     * Runtime FSM, to insert connectivity events
+     */
+    @Getter
+    private final Map<LogReplicationSession, CorfuLogReplicationRuntime> sessionToRuntimeFSM;
+
+    /**
+     * Remote cluster's clusterDescriptor
+     */
+    @Getter
+    private final Map<LogReplicationSession, ClusterDescriptor> sessionToRemoteClusterDescriptor;
+
+    /**
+     * The replicationManager
+     */
+    private final CorfuReplicationManager replicationManager;
```

Comment:
introduces a circular dependency because replicationManager also depends on the router to create the runtimes.

Reply:
I see that we cannot eliminate this dependence completely, in fact it existed in the prior LR also because the router generates events for the fsm.

But storing this variable can be avoided.  Replication Manager is used to get the session -> runtime map in the constructor.  This can be passed from SessionManager directly.  

I also put a comment where replicationManager.start() is invoked in startReplication().